README.md
# Personal Crypto Tracker
A simple tracker for your cryptocurrency investments. Input your deposits and crypto holdings in the CSV files, this app will display all the prices of the cryptos you own on the terminal and calculate the net value of your investments. It's always a good idea to keep track of your investments using a spreadsheet, this app encourages you as you will need to input your data into a CSV file and also serves to provide concise summary of your investments.

## Dependencies
App written in `python 3.10.1`, other versions may or may not work.
- requests
- bs4
- colorama
- google-currency

## Install and run app
1. Clone this repository
2. Open terminal/command line
3. Change directory `cd` to this app
4. Execute `python3 tracker.py`

## CSV formatting
Two CSV files are needed (`crypto.csv` and `deposits.csv`) and must be placed inside the `crypto/` directory. They must follow the required format below.

### crypto.csv
- The `Coin` column must must hold strings that allow access to `https://coinmarketcap.com/currencies/<coin>`.
- The `Symbol` column holds strings that are only for display.
- The `Holdings` column must be a numeric value that can be cast to `float`. Commas in numbers can be parsed (e.g., 1,000).

| Coin | Symbol | Holdings |
| -- | -- | -- |
| Bitcoin | BTC | 0.013 |
| Ethereum | ETH | 0.16 |

### deposits.csv
- The `Date` column is currently unused, its use may be implemented in the future.
- The `Amount` column must hold numeric values that can be cast to `float`. Commas in numbers can be parsed (e.g., 1,000). DO NOT include currency symbols (e.g., $).
- The `Currency` column must hold strings that abide to the ISO 4217 currency codes.

| Date | Amount | Currency |
| -- | -- | -- |
| Jan 1, 2022 | 1000 | USD |
| Jan 2, 2022 | 200 | EUR |

## Future plans
I am planning to add a GUI to this tracker for more data visualization.


CHANGELOG
v1.3.0
- Fixed inconsistent currency exchange by implemented the google_currency library

v1.2.0
- Fixed html scraper for currency rates
- Refactored code with Wallet class

v1.1.0
- Implemented local currency converter in place of the unreliable forex_python
- Implemented color coding on terminal for gains and losses

v1.0.0
- First release of terminal app


crypto.py
import re
import requests
from bs4 import BeautifulSoup
from locale import atof, Error as localeError
from colorama import Back, Style


class Crypto:
    """
    Crypto(coin, symbol, holding, goal_rate=None, currency="USD")
        Arguments:
        - coin (str): Name of cryptocurrency. Must be a working URL to
          https://coinmarketcap.com/currencies/<coin>
        - symbol (str): Short-form of coin
        - holding (float): Amount of coin holding
        - goal_rate (float) (default: None): Price goal for crypto
        - currency (str) (default: USD): Currency that goal_rate operates with.
          Must comply to ISO 4217 names.
    """

    header = "Rank\tCoin\tRates (USD)\t24hr Change\tHoldings (Coin)\t Holdings (USD)"
    url_prefix = "https://coinmarketcap.com/currencies/"
    rank_html_class = "namePill namePillPrimary"
    price_html_class = "priceValue"
    price_change_24h_up_html_class = "sc-15yy2pl-0 gEePkg"
    price_change_24h_down_html_class = "sc-15yy2pl-0 feeyND"

    def __init__(self, coin, symbol, holding, goal_rate=None, currency="USD"):
        self.coin = coin
        self.symbol = symbol
        self.holding = holding
        self.goal_rate = goal_rate
        self.currency = currency
        self.url = f"{self.url_prefix}{self.coin}"
        self.isUp = True
        self.color = Back.GREEN
        self.fetch_data()
        self.parse_rank()
        self.parse_price()
        self.parse_price_change_24h()
        self.update_holding()

    def fetch_data(self):
        """Fetch data from url"""
        try:
            self.page = requests.get(self.url)
        except requests.ConnectionError:
            raise Exception("Err#200: Network problem, connection error.")
        except requests.HTTPError as e:
            raise Exception(f"Err#201: HTTPError {e}.")
        except requests.Timeout as e:
            raise Exception(f"Err#202: Connection timeout. {e}")
        except requests.TooManyRedirects:
            raise Exception("Err#203: Too many redirects.")
        except Exception as e:
            raise Exception(f"{e}\nErr#299: Something went terribly wrong.")
        self.data = BeautifulSoup(self.page.content, "html.parser")

    def parse_rank(self):
        """Parse coin rank"""
        try:
            rank = self.data.find_all(class_=self.rank_html_class)
        except Exception as e:
            raise Exception(f"{e}\nErr#210: Could not parse rank.")
        if not rank:
            raise Exception("Err#210: Could not parse rank.")

        try:
            # Regex for looking between # and <
            self.rank = int(re.search(r"(?<=#)(.*?)(?=<)", str(rank)).group(0))
        except Exception:
            raise Exception("Err#211: This is why I hate regex.")

    def parse_price(self):
        """Parse current price"""
        try:
            price = self.data.find_all(class_=self.price_html_class)
        except Exception as e:
            raise Exception(f"{e}\nErr#221: Could not parse current price.")
        if not price:
            raise Exception("Err#221: Could not parse current price.")

        try:
            # Regex for looking between $ and <
            self.price = atof(re.search(r"(?<=\$)(.*?)(?=<)", str(price))
                              .group(0))
        except localeError:
            raise Exception("Err#223: Locale error")
        except Exception as e:
            raise Exception(f"{e}\nErr#222: This is why I hate regex.")

    def parse_price_change_24h(self):
        """Parse price change last 24-hour"""
        try:
            price_change_24h = self.data.find_all(
                class_=self.price_change_24h_up_html_class)
            self.isUp = True
            self.color = Back.GREEN
            if not price_change_24h:
                price_change_24h = self.data.find_all(
                    class_=self.price_change_24h_down_html_class)
                self.isUp = False
                self.color = Back.RED
        except Exception as e:
            raise Exception(f"{e}\nErr#231: Could not parse price change last 24-hour.")
        if not price_change_24h:
            raise Exception("Err#231: Could not parse price change last 24-hour.")

        try:
            # Regex for looking between $ and <
            self.price_change_24h = atof(re.search(
                r"(?<=span>)(.*?)(?=<\!)", str(price_change_24h[0])).group(0))
            if not self.isUp:
                self.price_change_24h *= -1
        except localeError:
            raise Exception("Err#233: Locale error")
        except Exception as e:
            raise Exception(f"{e}\nErr#232: This is why I hate regex.")

    def update_holding(self):
        """Convert holding in coin to fiat"""
        self.holding_in_fiat = self.holding * self.price

    def __repr__(self):
        return f"Crypto({self.coin}, {self.symbol}, {self.holding},\
                        goal_rate={self.goal_rate},\
                        currency={self.currency})"

    def __str__(self):
        return f"{self.color}{self.rank}\t{self.symbol}\t{self.price:<10}\t{self.price_change_24h}%\t\t{self.holding:<10}\t {self.holding_in_fiat:.7f}"


class Wallet:

    def __init__(self, wallet=[], deposits=[]):
        self.wallet = wallet
        self.deposits = deposits

        self.update_total_holdings()
        self.update_total_deposits()

    def add_coin(self, coin):
        self.wallet.append(coin)
        self.update_total_holdings()

    def add_deposit(self, deposit):
        self.deposits.append(deposit)
        self.update_total_deposits()

    def update_total_holdings(self):
        self.total_holdings = sum([cryp.holding_in_fiat for cryp in self.wallet])

    def update_total_deposits(self):
        self.total_deposits = sum([depos["Amount"] for depos in self.deposits])

    def update_rates(self):
        for cryp in self.wallet:
            cryp.fetch_data()
            cryp.parse_rank()
            cryp.parse_price()
            cryp.parse_price_change_24h()
            cryp.update_holding()
        self.update_total_holdings()


tracker.py
import os
import csv
import time
import json
from locale import setlocale, atof, LC_ALL
from crypto import Crypto, Wallet
from colorama import init, Back, Style
from google_currency import convert


def read_crypto_file(file_name):
    """
    Read crypto.csv and returns a list of crypto objects
    """
    print("Loading crypto...")

    try:
        with open(file_name, "r") as csv_file:
            csv_reader = csv.reader(csv_file, delimiter=",", quotechar="|")
            next(csv_reader, None)  # Skip header
            for row in csv_reader:
                w.add_coin(Crypto(row[0], row[1], atof(row[2])))
    except IOError:
        raise Exception("Err#100: CSV file for crypto not found.")
    except Exception as e:
        raise Exception(f"{e}\nErr#109: Something went terribly wrong.")


def read_deposits_file(file_name):
    """
    Read deposits.csv and returns a list deposits in dict
    """
    print("Loading deposits...")

    try:
        with open(file_name, "r") as csv_file:
            csv_reader = csv.reader(csv_file, delimiter=",")
            next(csv_reader, None)  # Skip header
            for row in csv_reader:
                w.add_deposit({"Date": row[0],
                               "Amount": atof(row[1]),
                               "Currency": row[2]})
    except IOError:
        raise Exception("Err#110: CSV file for crypto not found.")
    except Exception as e:
        raise Exception(f"{e}\nErr#119: Something went terribly wrong.")


def clear_console(): return os.system(
    'cls' if os.name in ('nt', 'dos') else 'clear')


def display_on_console():
    print(f"Total deposits: {w.total_deposits:.2f} USD\n")

    print(Crypto.header)
    print("-" * 90)
    w.update_rates()
    for cryp in w.wallet:
        print(cryp)

    try:
        print(f"\nTotal holdings\t| {w.total_holdings:.2f}")
        print(f"Total deposits\t| {w.total_deposits:.2f}")
        highlight_color = Back.GREEN if w.total_holdings >= w.total_deposits else Back.RED
        print(f"{highlight_color}Net (USD)\t| {w.total_holdings - w.total_deposits:.2f}")
        print(
            f"{highlight_color}Net (%)\t\t| {(w.total_holdings - w.total_deposits) / w.total_deposits * 100:.2f}%")
    except ZeroDivisionError:
        raise Exception("Err#120: 0 deposits, division by zero error.")
    except Exception as e:
        raise Exception(f"{e}\nErr#129: Something went terribly wrong.")

    print(
        f"\nLast updated time:\t{time.strftime('%H:%M:%S', time.localtime())}")


def main():
    read_crypto_file("crypto/crypto.csv")
    read_deposits_file("crypto/deposits.csv")

    print("Loading currency rates...")
    currency_pairs = {}
    for depos in w.deposits:
        currency_pair = depos["Currency"] + "USD"
        if currency_pair not in currency_pairs.keys():
            for i in range(5):
                currency_converter = json.loads(convert(depos["Currency"], "USD", 1))
                if currency_converter["converted"]:
                    break
                elif i == 4:
                    raise Exception(
                        f"Err#301: two possible errors here\n1. Bad currency code: {depos}\n2. API for converting currency failed 5 times in a row.")
            currency_pairs[currency_pair] = atof(currency_converter["amount"])
        depos["Amount"] *= currency_pairs[currency_pair]
        depos["Currency"] = "USD"
    w.update_total_deposits()

    start_time = time.time()
    while True:
        clear_console()
        display_on_console()
        time.sleep(180 - (time.time() - start_time) % 180)


if __name__ == "__main__":
    # os.chdir(os.path.dirname(__file__))
    try:
        setlocale(LC_ALL, "en_US.utf8")
    except Exception:
        try:
            setlocale(LC_ALL, "en_US.UTF-8")
        except Exception as e:
            raise Exception(f"{e}\nErr#000: locale error.")
    init(autoreset=True)
    clear_console()

    print("Initializing wallet...")
    w = Wallet()

    main()


currency.py
import re
import requests
from bs4 import BeautifulSoup
from locale import atof, Error as localeError


class CurrencyConverter:

    url_prefix = "https://www.marketwatch.com/investing/currency/"
    price_html_class = "intraday__price"

    def get_rate(self, from_curr, to_curr):
        """Fetch data from url"""
        try:
            page = requests.get(f"{self.url_prefix}{from_curr}{to_curr}")
        except requests.ConnectionError:
            print("Err#300: Network problem, connection error.")
        except requests.HTTPError as e:
            print(f"Err#301: HTTPError {e}.")
        except requests.Timeout as e:
            print(f"Err#302: Connection timeout. {e}")
        except requests.TooManyRedirects:
            print("Err#303: Too many redirects.")
        except Exception as e:
            print(f"{e}\nErr#399: Something went terribly wrong.")
        data = BeautifulSoup(page.content, "html.parser")

        price = self.parse_price(data)
        return price

    def parse_price(self, data):
        """Parse current price"""
        try:
            price = data.find_all(class_=self.price_html_class)
        except Exception as e:
            print(f"{e}\nErr#321: Could not parse current price.")
        if not price:
            raise Exception("Err#321: Could not parse current price.")

        try:
            # Regex for looking between value"> and </span
            price = atof(re.search(r"(?<=>)(.*?)(?=</bg-quote)", str(price))
                         .group(0))
        except localeError:
            print("Err#323: Locale error")
        except Exception as e:
            print(f"{e}\nErr#322: This is why I hate regex.")

        return price


app.py
import os
import sys
import csv
import time
import json
import threading
from locale import setlocale, atof, LC_ALL

from PyQt5.QtCore import Qt
from PyQt5.QtWidgets import (QApplication, QTabWidget, QWidget, QVBoxLayout,
                             QTableWidget, QAbstractItemView, QAbstractScrollArea, QDialog, QLabel, QDesktopWidget)

from crypto import Crypto, Wallet


class Loading(QDialog):

    def __init__(self):
        super().__init__()
        self.setWindowTitle("Loading...")
        self.setFixedSize(300, 100)
        self.center()
        self.setWindowModality(Qt.ApplicationModal)
        self.setWindowFlag(Qt.WindowCloseButtonHint, False)

        vbox = QVBoxLayout()
        self.lbl = QLabel("Initializing wallet.")
        self.lbl.setAlignment(Qt.AlignCenter)
        vbox.addWidget(self.lbl)
        self.setLayout(vbox)
        self.show()

        self.allowed_to_close = False

        thread = threading.Thread(target=self.loop_text)
        thread.start()

    def center(self):
        qr = self.frameGeometry()
        cp = QDesktopWidget().availableGeometry().center()
        qr.moveCenter(cp)
        self.move(qr.topLeft())

    def loop_text(self):
        while not self.allowed_to_close:
            time.sleep(1)
            if str(self.lbl.text()).count('.') == 1:
                self.lbl.setText("Initializing wallet..")
            elif str(self.lbl.text()).count('.') == 2:
                self.lbl.setText("Initializing wallet...")
            elif str(self.lbl.text()).count('.') == 3:
                self.lbl.setText("Initializing wallet.")

    def closeEvent(self, evnt):
        if self.allowed_to_close:
            super().closeEvent(evnt)
        else:
            evnt.ignore()


class Window(QTabWidget):

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setWindowTitle("Personal Crypto Tracker")
        self.setGeometry(50, 50, 600, 400)

        # TODO: progress bar
        self.init_loading()

        self.tab1 = QWidget()
        self.tab2 = QWidget()
        self.tab3 = QWidget()

        self.addTab(self.tab1, "Crypto Wallet")
        self.addTab(self.tab2, "Fiat Deposits and Withdrawal")
        self.addTab(self.tab3, "Trade History")
        self.tab1UI()
        self.tab2UI()
        self.tab3UI()

        self.show()

    def init_loading(self):
        self.dialog = Loading()

        thread = threading.Thread(target=self.read_wallet_thread)
        thread.start()

    def tab1UI(self):
        vbox = QVBoxLayout()

        self.tab1.setLayout(vbox)

    def tab2UI(self):
        vbox = QVBoxLayout()
        table = QTableWidget(5, 3)
        table.setEditTriggers(QAbstractItemView.NoEditTriggers)
        table.setSizeAdjustPolicy(QAbstractScrollArea.AdjustToContents)
        table.resizeColumnsToContents()
        table.setHorizontalHeaderLabels(["Date", "Amount", "Currency"])
        vbox.addWidget(table)

        self.tab2.setLayout(vbox)

    def tab3UI(self):
        vbox = QVBoxLayout()

        self.tab3.setLayout(vbox)

    def read_wallet_thread(self):
        self.w = Wallet()
        self.read_crypto_file("crypto/crypto.csv")
        self.read_deposits_file("crypto/deposits.csv")
        print("Done")
        self.dialog.allowed_to_close = True
        self.dialog.close()

    def read_crypto_file(self, file_name):
        """
        Read crypto.csv and returns a list of crypto objects
        """
        print("Loading crypto...")

        try:
            with open(file_name, "r") as csv_file:
                csv_reader = csv.reader(csv_file, delimiter=",", quotechar="|")
                next(csv_reader, None)  # Skip header
                for row in csv_reader:
                    self.w.add_coin(Crypto(row[0], row[1], atof(row[2])))
        except IOError:
            raise Exception("Err#100: CSV file for crypto not found.")
        except Exception as e:
            raise Exception(f"{e}\nErr#109: Something went terribly wrong.")

    def read_deposits_file(self, file_name):
        """
        Read deposits.csv and returns a list deposits in dict
        """
        print("Loading deposits...")

        try:
            with open(file_name, "r") as csv_file:
                csv_reader = csv.reader(csv_file, delimiter=",")
                next(csv_reader, None)  # Skip header
                for row in csv_reader:
                    self.w.add_deposit({"Date": row[0],
                                        "Amount": atof(row[1]),
                                        "Currency": row[2]})
        except IOError:
            raise Exception("Err#110: CSV file for crypto not found.")
        except Exception as e:
            raise Exception(f"{e}\nErr#119: Something went terribly wrong.")


def main():
    try:
        setlocale(LC_ALL, "en_US.utf8")
    except Exception:
        try:
            setlocale(LC_ALL, "en_US.UTF-8")
        except Exception as e:
            raise Exception(f"{e}\nErr#000: locale error.")

    app = QApplication(sys.argv)
    window = Window()
    sys.exit(app.exec_())


if __name__ == "__main__":
    main()
